### DISCUSSION TITLE
Energy calculation while simulation

### DISCUSSION THREAD

**Discussion by GenB31415** [zwicker-group/py-pde#103] [Q&A]
While simulating the Brusselator (with spatial coupling) with u(x, y, t) and v(x, y, t), I have to calculate the energy 
W(t) = integral ((u ^ 2 + v ^ 2) dxdy) (or other analogous quantities) at times t < tFin. How do I access the local values u(x, y, t) and v(x, y, t) during the simulation (run-time) to have W(t) for the later use? Is there a general technique? Thanks.

**Comment by david-zwicker** [comment]
The most straight-forward way would probably to use a [`DataTracker`](https://py-pde.readthedocs.io/en/latest/packages/pde.trackers.html#pde.trackers.trackers.DataTracker) to call a custom function periodically. 
Here is a modification of the example that shows the ideas in action:

```
from pde import PDE, FieldCollection, DataTracker, ScalarField, UnitGrid
import matplotlib.pyplot as plt

# define the PDE
a, b = 1, 3
d0, d1 = 1, 0.1
eq = PDE(
    {
        "u": f"{d0} * laplace(u) + {a} - ({b} + 1) * u + u**2 * v",
        "v": f"{d1} * laplace(v) + {b} * u - u**2 * v",
    }
)

# initialize state
grid = UnitGrid([64, 64])
u = ScalarField(grid, a, label="Field $u$")
v = b / a + 0.1 * ScalarField.random_normal(grid, label="Field $v$")
state = FieldCollection([u, v])


def get_energy(data):
    u, v = data
    return (u**2 + v**2).integral

# simulate the pde
data_tracker = DataTracker(get_energy, interval=1)
eq.solve(state, t_range=20, dt=1e-3, tracker=['progress', data_tracker])

plt.plot(data_tracker.times, data_tracker.data)
```

**Reply by GenB31415** [reply]
For my case it works fine. Thank you!

**Comment by GenB31415** [comment]
In my case this solution works fine after minor corrections
1. from pde import PDE, FieldCollection, PlotTracker, ScalarField, UnitGrid, DataTracker <<-- add DataTracker 2. pde.DataTracker(  <<--- DataTracker(  omit  pde.
Thank you!

    On Tuesday, May 4, 2021, 12:50:55 PM CDT, David Zwicker ***@***.***> wrote:  
 
 


The most straight-forward way would probably to use a DataTracker to call a custom function periodically.
Here is a modification of the example that shows the ideas in action:
from pde import PDE, FieldCollection, PlotTracker, ScalarField, UnitGrid
import matplotlib.pyplot as plt

# define the PDE
a, b = 1, 3
d0, d1 = 1, 0.1
eq = PDE(
    {
        "u": f"{d0} * laplace(u) + {a} - ({b} + 1) * u + u**2 * v",
        "v": f"{d1} * laplace(v) + {b} * u - u**2 * v",
    }
)

# initialize state
grid = UnitGrid([64, 64])
u = ScalarField(grid, a, label="Field $u$")
v = b / a + 0.1 * ScalarField.random_normal(grid, label="Field $v$")
state = FieldCollection([u, v])


def get_energy(data):
    u, v = data
    return (u**2 + v**2).integral

# simulate the pde
data_tracker = pde.DataTracker(get_energy, interval=1)
eq.solve(state, t_range=20, dt=1e-3, tracker=['progress', data_tracker])

plt.plot(data_tracker.times, data_tracker.data)

**Reply by GenB31415** [reply]
I'd like to extend this solution to see both plots: 2D and the energy in the same time. The idea is to use the trackers as in the examples. My code is below

from pde import PDE, FieldCollection, PlotTracker, ScalarField, UnitGrid, DataTracker
import matplotlib.pyplot as plt

# define the PDE
a, b = 1, 3
d0, d1 = 1, 0.1
eq = PDE(
    {
        "u": f"{d0} * laplace(u) + {a} - ({b} + 1) * u + u**2 * v",
        "v": f"{d1} * laplace(v) + {b} * u - u**2 * v",
    }
)

# initialize state
grid = UnitGrid([64, 64])
u = ScalarField(grid, a, label="Field $u$")
v = b / a + 0.1 * ScalarField.random_normal(grid, label="Field $v$")
state = FieldCollection([u, v])
def get_energy(data):
    u, v = data
    return (u**2 + v**2).integral

showThis=False
#showThis=True  # it is interesting that dependently of comment/uncomment this string I see only one of plots!
trackers = [
    "progress",  # show progress bar during simulation    
    #storage.tracker(interval=1),  # store data every simulation time unit
    PlotTracker(show=showThis),  # show images during simulation    
    DataTracker(get_energy), 
]
eq.solve(state, t_range=5, dt=1e-3, tracker=trackers)
data_tracker = trackers[2]
print(data_tracker.times)
plt.plot(data_tracker.times, data_tracker.data, marker="o" )
plt.grid(True); plt.xlabel("Time"); plt.ylabel("Energy"); plt.title("Plot")
plt.show()

My question is as follows: dependently of comment/uncomment of the string (#showThis=True) I see one of the plots! But how to see both plots of the field and the energy in the same time? Thank you.

**Reply by david-zwicker** [reply]
This is not implemented. If you want to implement it yourself, you can hook into the simulation using a [`CallbackTracker`](https://py-pde.readthedocs.io/en/latest/packages/pde.trackers.html#pde.trackers.trackers.CallbackTracker)

**Reply by GenB31415** [reply]
Thanks
